### ISSUE TITLE
[Perf][Qwen3-Omni] Drop bulk .cpu().tolist() embedding conversions from the talker request-build path

### ISSUE THREAD

**Issue by Hayden727** [sgl-project/sglang-omni#484]
## Motivation

In `sglang-omni`, the talker stage's request-build path currently demotes an embedding tensor that is already GPU-resident and ready to feed the backend into a Python `list[list[float]]`. This demotion is a **critical-path tax on talker TTFT**, for two reasons:

1. It happens before talker enters prefill, so **every request pays it once**.
2. It simultaneously incurs four costs: a GPU→CPU copy, a massive Python-object allocation burst, transient memory inflation, and an immediately following CPU→GPU round-trip when upstream SGLang converts it back into a tensor.

In other words, we perform two fully reversed tensor↔list conversions back-to-back, with nothing in between consuming the list form for its own sake.

### 1. Two tensor → list conversions on the talker request-build path

`sglang_omni/models/qwen3_omni/request_builders.py:467-481`

```python
if talker_input_embeds is not None:
    input_embeds = talker_input_embeds.float().cpu().tolist()
    input_ids_tensor = torch.as_tensor(talker_input_ids, dtype=torch.long)
    input_ids_list = input_ids_tensor.tolist()
    seq_len = len(input_ids_list)
else:
    # thinker_hidden_states: [seq_len, thinker_hidden_size]
    seq_len = thinker_hidden_states.shape[0]
    input_ids_list = [codec_bos_id] * seq_len
    input_ids_tensor = torch.tensor(input_ids_list, dtype=torch.long)
    # Convert hidden states to list-of-lists for Req.input_embeds
    input_embeds = thinker_hidden_states.float().cpu().tolist()
```

The result is written directly into `Req.input_embeds`:

`sglang_omni/models/qwen3_omni/request_builders.py:497-505`

```python
req = Req(
    rid=rid,
    origin_input_text="",
    origin_input_ids=input_ids_list,
    sampling_params=sampling_params,
    input_embeds=input_embeds,
    eos_token_ids={int(codec_eos_id)} if codec_eos_id is not None else None,
    vocab_size=codec_vocab_size,
)
```

`talker_input_embeds` originates from `prefill["input_embeds"]` returned by `talker_prefill.build_prompt_prefill(...)`, which `build_prefill_input(...)` assembles on the GPU in the talker's dtype (`sglang_omni/models/qwen3_omni/components/talker_prefill.py:209-236`). Put differently, the source tensor on this path is already **device-resident, dtype-correct, and layout-correct**; there is no technical reason it must be demoted to a Python list.

### 2. Upstream SGLang immediately converts it back to a tensor

`/sglang/srt/managers/schedule_batch.py:594`

```python
input_embeds: Optional[List[List[float]]] = None,
```

`schedule_batch.py:1778-1784` — during `prepare_for_extend`, the list is sliced by `extend_input_len`:

```python
if req.input_embeds is not None:
    input_embeds.extend(
        req.input_embeds[pre_len : pre_len + req.extend_input_len]
    )
```

`schedule_batch.py:1917-1921` — and finally lifted back to a tensor via `torch.tensor(...).to(device)`:

```python
self.input_embeds = (
    torch.tensor(input_embeds, pin_memory=_pin).to(
        ...
    )
)
```

So the actual data flow today is:

```
GPU tensor [seq_len, hidden] (talker dtype, contiguous)
  → .float()                        # upcast, new tensor allocation
  → .cpu()                          # GPU→CPU copy
  → .tolist()                       # creates seq_len * hidden Python float objects
  → stored in Req.input_embeds (list[list[float]])
  → upstream prepare_for_extend slices it row-by-row
  → torch.tensor(...).pin_memory().to(device)   # CPU→GPU copy back
```

Every Python object and every CPU buffer in the middle is one-shot, pure overhead. On long prompts at `seq_len * hidden_size` order of magnitude (e.g., 4096 × 2048), a single build creates roughly 8M Python float objects.

### 3. Upstream prefill already forces full (non-chunked) prefill for this request

`sglang_omni/scheduling/sglang_backend/prefill.py:44-76` — `_needs_full_prefill(...)` identifies talker-style requests via the `req._input_embeds_are_projected` flag and explicitly sets `disable_chunking = True`. This means the talker request is **dropped into the prefill batch as one big extend in a single step**; the list-of-lists cost is not amortized across multiple chunks. The full list-ification cost fires at exactly the moment the request enters the waiting queue.

## Proposal

Keep the talker request-build path tensor-native: **avoid any unnecessary `.float()` / `.cpu()` / `.tolist()`**.

There are two implementation directions that can be staged separately:

### Plan A (in scope for this issue, must-do) — shrink the visible surface of list-ification

Without touching the upstream SGLang contract `Req.input_embeds: Optional[List[List[float]]]`:

1. Stop performing `.float().cpu().tolist()` explicitly inside `build_sglang_talker_request(...)`. Defer the Python-list demotion to the exact line that writes `Req.input_embeds`, and avoid the unnecessary `.float()` upcast — upstream's batch-tensor construction already runs `torch.tensor(input_embeds, pin_memory=_pin).to(device, dtype=...)`, so the dtype/device conversion is going to happen there anyway; doing fp32 ahead of time is wasted work.
2. Evaluate whether `input_ids_list` can be passed as a 1D tensor (`torch.as_tensor(...)`) instead of going through `.tolist()`. This depends on the upstream `Req.origin_input_ids` contract — if it accepts a sequence-of-ints, we can avoid one Python list materialization.
3. Add a microbenchmark for this path to the talker-request-build test surface, isolating `build_sglang_talker_request` wall-clock and GC pressure at different `seq_len`s.

Plan A is the minimum viable variant: **"don't change the contract, just stop paying for things we don't need."** Even without any upstream change, this alone should produce measurable TTFT wins on long prompts.

### Plan B (out of scope for this issue; may become a follow-up RFC) — push tensor-backed `Req.input_embeds` upstream

The deeper fix is to let upstream `Req.input_embeds` accept `torch.Tensor` (device-resident, shape `[seq_len, hidden]`) and have `prepare_for_extend` slice tensor rows directly rather than list rows. That path requires:

- An upstream PR that touches the `Req` type signature and the slice/concat path inside `ScheduleBatch.prepare_for_extend`.
- A full audit of every downstream model writing into `req.input_embeds` (not just talker).
- Confirmation that the current `pin_memory` + `to(device)` movement semantics are preserved on the tensor path.

Plan B should **not** be attempted inside this issue. Once Plan A demonstrates the TTFT win and quantifies the list-ification cost, that data becomes the evidence base for a standalone upstream RFC.

## Invariants (must hold)

- Talker output (audio quality, WER, waveform parity) must not regress.
- `Req.input_embeds_are_projected` semantics stay unchanged (talker requests still take the disable-chunked-prefill path).
- Downstream consumers of `input_ids_tensor` / `talker_input_ids` — `_compute_mrope_positions`, `MultimodalInputs`, `multimodal_mask`, etc. — must remain side-effect-free.

## Non-Goals

- No rewrite of `pending_text_queue` (per-row CPU queue → tensor-backed) in this issue — that is a separate follow-up.
- No Code2Wav incremental-buffer changes in this issue.
- No scheduler / outbox wakeup changes in this issue.
- No change to the upstream SGLang `Req.input_embeds` type signature in this issue — Plan B stays as a separate RFC.

## Relation to existing work

- Orthogonal to #472 (topology-aware transport): #472 is about **how data moves** between stages; this issue is about **not making the data take a wasteful self-loop the moment it enters talker**.

## Acceptance Metrics

- `build_sglang_talker_request(...)` wall-clock drops noticeably on long prompts (e.g., `seq_len ≥ 2048`), quantified via a microbenchmark.
- Python object allocation during talker request-build drops measurably (estimable via `tracemalloc` or `gc` hooks).
- End-to-end audio TTFT P50 / P95 on SeedTTS / realtime workloads improves measurably.
- WER / waveform-parity tests do not regress.

## Open Questions

1. Are there non-talker producers of `Req.input_embeds` that depend on `list[list[float]]`? This determines the scope of any future Plan B RFC.
2. Does `Req.origin_input_ids` safely accept non-list inputs? Needs verification against the upstream contract.
3. What is the dtype and device of `prefill["input_embeds"]` on the talker prefill path today? If it is already in the talker's dtype on the talker's device, the `.float()` in Plan A step 1 can be removed outright; otherwise we need to confirm the expected `pin_memory + .to(device, dtype=...)` movement cost on the upstream side.
4. Should this issue add a standalone CI microbenchmark (independent of a full server) as a long-term regression guard?

**Comment by zhaochenyang20** [comment]
@SandyLuXY

**Comment by yxs** [comment]
@SandyLuXY 

Same 30 s steady-state capture as #485:

- `cudaMemcpyAsync`: 1.79 s / 30 s, 4,669 instances. A meaningful share is the prefill-time `.float().cpu()` + the upstream `torch.tensor(input_embeds).pin_memory().to(device, dtype=...)`round-trip.
- Total Python object allocation per request build is bounded by `seq_len × hidden` floats. For a typical seq_len ~2 k that is ~4 M PyObject* allocations on the critical path.

Plan A is clean — defer `.tolist()` to the Req write line and drop the `.float()` upcast. A 1R on top of #487 with the change applied (combined with the #485 patch) didn't move bulk `latency_mean` (within noise), so TTFT P50 / P95 is the right validation surface.
